syntax = "proto3";

service Niraapad {

  // Used when the client-server connection starts
  // i.e., when the NiraapadClientHelper class is initialized
  rpc InitializeConnection(InitializeConnectionReq) returns (InitializeConnectionResp) {}

  // Used when the client-server connection stops
  // i.e., when the NiraapadClientHelper class is deleted
  rpc DeleteConnection(DeleteConnectionReq) returns (DeleteConnectionResp) {}

  // Used for all static and class methods
  rpc StaticMethod(StaticMethodReq) returns (StaticMethodResp) {}

  // Used for all static and class variable read access
  rpc StaticGetter(StaticGetterReq) returns (StaticGetterResp) {}

  // Used for all static and class variable write access
  rpc StaticSetter(StaticSetterReq) returns (StaticSetterResp) {}

  // Used for the __init__ method
  rpc Initialize(InitializeReq) returns (InitializeResp) {}

  // Used for the __del__ method
  rpc Uninitialize(UninitializeReq) returns (UninitializeResp) {}

  // Used for property getters
  rpc GenericGetter(GenericGetterReq) returns (GenericGetterResp) {}

  // Used for property setters
  rpc GenericSetter(GenericSetterReq) returns (GenericSetterResp) {}

  // Used for the remaining methods
  rpc GenericMethod(GenericMethodReq) returns (GenericMethodResp) {}

  // Used for sending traces (can be a batch or a single trace)
  rpc BatchedTrace(BatchedTraceMsg) returns (EmptyMsg) {}

  // Load the trace file for simulation / replay
  rpc LoadTrace(LoadTraceReq) returns (LoadTraceResp) {}
}

message CommandProfile {
  int32 mo = 1;
  float exec_time_sec = 2; 
}

message TimeProfile {
  uint64 id = 1;
  string arrival_time = 2;
  string departure_time = 3;
}

message StartServerTraceMsg {
}

message StopServerTraceMsg {
}

message InitializeConnectionReq {
}

message InitializeConnectionResp {
  bytes exception = 1;
}

message InitializeConnectionTraceMsg {
  InitializeConnectionReq req = 1;
  InitializeConnectionResp resp = 2;
}

message DeleteConnectionReq {
}

message DeleteConnectionResp {
  bytes exception = 1;  
}

message DeleteConnectionTraceMsg {
  DeleteConnectionReq req = 1;
  DeleteConnectionResp resp = 2;
}

message StaticMethodReq {
  string arrival_time = 1;
  string backend_type = 2;
  string method_name = 3;
  bytes args = 4;
  bytes kwargs = 5;
  uint64 id = 6;
  repeated TimeProfile time_profiles = 7;
}

message StaticMethodResp {
  bytes exception = 1;
  bytes resp = 2;
}

message StaticMethodTraceMsg {
  StaticMethodReq req = 1;
  StaticMethodResp resp = 2;
  CommandProfile profile = 3;
}

message StaticGetterReq {
  string arrival_time = 1;
  string backend_type = 2;
  string property_name = 3;
  uint64 id = 4;
  repeated TimeProfile time_profiles = 5;
}

message StaticGetterResp {
  bytes exception = 1;
  bytes resp = 2;
}

message StaticGetterTraceMsg {
  StaticGetterReq req = 1;
  StaticGetterResp resp = 2;
  CommandProfile profile = 3;
}

message StaticSetterReq {
  string arrival_time = 1;
  string backend_type = 2;
  string property_name = 3;
  bytes value = 4;
  uint64 id = 5;
  repeated TimeProfile time_profiles = 6;
}

message StaticSetterResp {
  bytes exception = 1;
}

message StaticSetterTraceMsg {
  StaticSetterReq req = 1;
  StaticSetterResp resp = 2;
  CommandProfile profile = 3;
}

message InitializeReq {
  string arrival_time = 1;
  string backend_type = 2;
  int32 backend_instance_id = 3;
  bytes args = 4;
  bytes kwargs = 5;
  bytes stacktrace = 6;
  uint64 id = 7;
  repeated TimeProfile time_profiles = 8;
}

message InitializeResp {
  bytes exception = 1;
}

message InitializeTraceMsg {
  InitializeReq req = 1;
  InitializeResp resp = 2;
  CommandProfile profile = 3;
}

message UninitializeReq {
  string arrival_time = 1;
  string backend_type = 2;
  int32 backend_instance_id = 3;
  uint64 id = 4;
  repeated TimeProfile time_profiles = 5;
}

message UninitializeResp {
  bytes exception = 1;
}

message UninitializeTraceMsg {
  InitializeReq req = 1;
  InitializeResp resp = 2;
  CommandProfile profile = 3;
}

message GenericGetterReq {
  string arrival_time = 1;
  string backend_type = 2;
  int32 backend_instance_id = 3;
  string property_name = 4;
  uint64 id = 5;
  repeated TimeProfile time_profiles = 6;
}

message GenericGetterResp {
  bytes exception = 1;
  bytes resp = 2;
}

message GenericGetterTraceMsg {
  GenericGetterReq req = 1;
  GenericGetterResp resp = 2;
  CommandProfile profile = 3;
}

message GenericSetterReq {
  string arrival_time = 1;
  string backend_type = 2;
  int32 backend_instance_id = 3;
  string property_name = 4;
  bytes value = 5;
  uint64 id = 6;
  repeated TimeProfile time_profiles = 7;
}

message GenericSetterResp {
  bytes exception = 1;
}

message GenericSetterTraceMsg {
  GenericSetterReq req = 1;
  GenericSetterResp resp = 2;
  CommandProfile profile = 3;
}

message GenericMethodReq {
  string arrival_time = 1;
  string backend_type = 2;
  int32 backend_instance_id = 3;
  string method_name = 4;
  bytes args = 5;
  bytes kwargs = 6;
  uint64 id = 7;
  repeated TimeProfile time_profiles = 8;
}

message GenericMethodResp {
  bytes exception = 1;
  bytes resp = 2;
}

message GenericMethodTraceMsg {
  GenericMethodReq req = 1;
  GenericMethodResp resp = 2;
  CommandProfile profile = 3;
}

// Not using repeated for trace_msgs since each msg
// may be of a different type
message BatchedTraceMsg {
  bytes trace_msgs = 1;
}

// Trace header helps identify the trace length
message TraceHeader {
    // number of bytes in the following trace metadata
    int32 metadata_size = 1;
}

// Trace metadata helps identify the trace message type and its size
message TraceMetadata {
  string timestamp = 1;
  string trace_msg_type = 2;

  // number of bytes in the following trace msg
  // i.e., any message type above ending in "TraceMsg"
  int32 trace_msg_size = 3;
}

// Since gRPC does not allow defining RPC with void params / return values
message EmptyMsg {
}

message LoadTraceReq {
  string trace_file = 1;
}

message LoadTraceResp {
  bool status = 1;
}